File: scripts/algorithms/M/Minimum Moves to Move a Box to Their Target Location/Minimum Moves to Move a Box to Their Target Location.java
// Runtime: 51 ms (Top 51.89%) | Memory: 54.4 MB (Top 47.17%)
/**
  Finds Initial State (State consists of shopkeeper & box locations + # of boxMoves)
  Uses BFS/A* Algorithm to visit valid transition states
  Note: The min heuristic here is # of boxMoves + manHattanDistance between box & target locations
*/
class Solution {
    private int targetRow;
    private int targetCol;
    private char[][] grid;
    private static int[][] DIRS = {
            {1,0}, //Down
            {-1,0},//Up
            {0,1}, //Right
            {0,-1} //Left
    };

    /**
      State holds shopkeeper and box location, as well as how many times the box has been pushed
    */
    class State implements Comparable<State>{
        int personRow;
        int personCol;
        int boxRow;
        int boxCol;
        int boxPushes;

        public State(int personRow, int personCol, int boxRow, int boxCol, int boxPushes){
            this.personRow = personRow;
            this.personCol = personCol;
            this.boxRow = boxRow;
            this.boxCol = boxCol;
            this.boxPushes = boxPushes;
        }

        // Override equals - used along with hashcode when we have visited HashSet
        public boolean equals(Object o){
            State other = (State) o;
            return
            this.personRow == other.personRow &&
            this.personCol == other.personCol &&
            this.boxRow == other.boxRow &&
            this.boxCol == other.boxCol;
        }

        // Override the hashCode - Note: it's okay for this to have collisions
        // But it won't due to the problem constraint that there is a bound on NxM dimensions
        public int hashCode(){
            return personRow *10_000 + personCol * 1_000 + boxRow * 100 + boxCol;
        }

        // Override to string method - helpful in debugging state.
        public String toString(){
            return "ShopKeeper:{row:"+personRow+", col:"+personCol+"}" +
                   "Box:{row:"+boxRow+", col:"+boxCol+"}";
        }

        // Implement comparable interface such that we return the state that
        // has the possibility of lowest distance using box push count + Manhattan distance
        public int compareTo(State other){
            int minDistanceThis = this.boxPushes + distanceToTarget(this);
            int minDistanceOther = other.boxPushes + distanceToTarget(other);
            return Integer.compare(minDistanceThis, minDistanceOther);
        }

    }

    // Calculates Manhattan distance
    private int distanceToTarget(State state){
        int yDiff = Math.abs(state.boxCol - targetCol);
        int xDiff = Math.abs(state.boxRow - targetRow);
        return yDiff + xDiff;
    }

    /**
       Given a state, compare box location to target location to determine if it is
       a solution state.
    */
    private boolean isSolutionState(State state){
        return state.boxRow == targetRow && state.boxCol == targetCol;
    }

    /**
        Given a state, finds all valid transition states.
        This is accomplished by moving the ShopKeeper in all 4 directions and validate
        - Next ShopKeeper location is in bounds and is not a wall

        We have additional logic for when the next shopkeeper location is the box location:
        - Get next box location, by pushing the same direction, again validate that
          the next box location is in bounds, and is not a wall.

        If it's a valid transition, create the new state with the new shop keeper location
        and if the box moved, the new box location (also increment the number of box moves).

    **/
    private List<State> getNeighbors(State state){

        int personRow = state.personRow;
        int personCol = state.personCol;
        int boxRow = state.boxRow;
        int boxCol = state.boxCol;

        List<State> states = new ArrayList<>();
        for(int[] dir : DIRS){
            int rowMove = dir[0];
            int colMove = dir[1];
            int personRowNew = personRow + rowMove;
            int personColNew = personCol + colMove;
            // Shopkeeper cannot move into wall or go out of bounds skip to next direction
            if(!inBounds(personRowNew, personColNew) ||
               isWall(personRowNew, personColNew)){
                continue;
            }
            // Whether or not person will collide with box
            boolean willPushBox = personRowNew == boxRow && personColNew == boxCol;

            if(willPushBox){
                int boxRowNew = boxRow + rowMove;
                int boxColNew = boxCol + colMove;
                // Validate box can be pushed - if so push box and add to neighbor states
                if(inBounds(boxRowNew, boxColNew) &&
                   !isWall(boxRowNew, boxColNew)){
                   states.add(new State(personRowNew, personColNew, boxRowNew, boxColNew, state.boxPushes + 1));
                }
            } else {
                //Shop keeper moved, but not box
                states.add(new State(personRowNew, personColNew, boxRow, boxCol, state.boxPushes));
            }
        }
        return states;

    }

    // Given row/col, return whether it is wall
    private boolean isWall(int row, int col){
        char cell = grid[row][col];
        return cell == '#';
    }

    // Given row/col return whether is inBounds
    private boolean inBounds(int row, int col){
        int rows = grid.length;
        int cols = grid[0].length;
        if(row < 0 || col < 0 || row > rows-1 || col > cols-1){
            return false;
        }
        return true;
    }

    /**
      Returns initial state. Also finds and stores the target location.
    */
    private State getInitialState(){

        int shopKeeperRow=0;
        int shopKeeperCol=0;

        int boxRow = 0;
        int boxCol = 0;

        for(int r=0; r<grid.length; r++){
            char[] row = grid[r];
            for(int c=0; c<row.length; c++){
                char cell = grid[r][c];
                if(cell == 'T'){
                    this.targetRow = r;
                    this.targetCol = c;
                }
                else if(cell == 'B'){
                    boxRow = r;
                    boxCol = c;
                } else if(cell == 'S'){
                    shopKeeperRow = r;
                    shopKeeperCol = c;
                }
            }
        }

        return new State(shopKeeperRow, shopKeeperCol, boxRow, boxCol, 0);
    }

    public int minPushBox(char[][] grid) {
        this.grid = grid;

        State initialState = getInitialState();
        Queue<State> queue = new PriorityQueue<>();
        Set<State> visited = new HashSet<>();
        queue.offer(initialState);

        // Explore every state using BSF and keep track of the best solution
        while(!queue.isEmpty()){
            State state = queue.poll();
            if(visited.contains(state)){
                continue;
            }
            visited.add(state);
            /*
              Note: the reason we can return the first solution state we find, is because we are
              using priority queue with minDistance heuristic which means the first solution we find
              is guaranteed to be the optimal solution - (A* Algorithm)
            */
            if(isSolutionState(state)){
                return state.boxPushes;
            }
            for(State neighbor : getNeighbors(state)){
                if(!visited.contains(neighbor)){
                    queue.offer(neighbor);
                }
            };
        }
        // No solution - return -1
        return -1;
    }

}

File: scripts/algorithms/M/Minimum Moves to Move a Box to Their Target Location/Minimum Moves to Move a Box to Their Target Location.js
// Runtime: 252 ms (Top 66.67%) | Memory: 50.7 MB (Top 73.33%)
/**
 * @param {character[][]} grid
 * @return {number}
 */
var minPushBox = function(grid) {
    // common info & utils
    const m = grid.length
    const n = grid[0].length
    const dirs = [[1, 0], [-1, 0], [0, 1], [0, -1]]
    const add = (a, b) => [a[0] + b[0], a[1] + b[1]]
    const equals = (a, b) => a[0] === b[0] && a[1] === b[1]
    const validate = ([x, y]) => x >= 0 && y >= 0 && x < m && y < n
    const getKey = ([x, y]) => x * n + y

    // find all player, ball, target, and free cells
    const init = () => {
        let player
        let ball
        let target
        const freeSet = new Set()
        for (let i = 0; i < m; i++) {
            for (let j = 0; j < n; j++) {
                if (grid[i][j] === '.') {
                    freeSet.add(getKey([i, j]))
                } else if (grid[i][j] === 'S') {
                    freeSet.add(getKey([i, j]))
                    player = [i, j]
                } else if (grid[i][j] === 'B') {
                    freeSet.add(getKey([i, j]))
                    ball = [i, j]
                } else if (grid[i][j] === 'T') {
                    freeSet.add(getKey([i, j]))
                    target = [i, j]
                }
            }
        }
        return { player, ball, target, freeSet }
    }
    const { player, ball, target, freeSet } = init()
    const targetKey = getKey(target)

    // detect whether two cells are connected
    const connCache = new Map() // [x,y,x2,y2,x3,y3] => boolean
    const getConnKey = (a, b, ball) => {
        if (
            a[0] > b[0] ||
            a[0] === b[0] && a[1] > b[1]
        ) {
            [a, b] = [b, a]
        }
        return [a[0], a[1], b[0], b[1], ball[0], ball[1]].join(',')
    }
    const isConnected = (a, b, ball) => {
        if (a[0] === b[0] && a[1] === b[1]) {
            return true
        }
        const connKey = getConnKey(a, b, ball)
        if (connCache.has(connKey)) {
            return connCache.get(connKey)
        }
        const bKey = getKey(b)
        const ballKey = getKey(ball)
        const visited = new Array(m * n).fill(false)
        visited[getKey(a)] = true
        const queue = [a]
        while (queue.length) {
            const cell = queue.shift()
            for (let i = 0; i < dirs.length; i++) {
                const next = add(cell, dirs[i])
                if (validate(next)) {
                    const nextKey = getKey(next)
                    if (nextKey === bKey) {
                        connCache.set(connKey, true)
                        return true
                    }
                    if (
                        freeSet.has(nextKey) &&
                        nextKey !== ballKey &&
                        !visited[nextKey]
                    ) {
                        visited[nextKey] = true
                        queue.push(next)
                    }
                }
            }
        }
        connCache.set(connKey, false)
        return false
    }

    // solve the game
    const getStateKey = ([x, y], [xx, yy]) => [x, y, xx, yy].join(',') // ball, player
    const stateCache = new Set() // Set<stateKey>
    let queue = [[ball, player]]
    let count = 1
    while (queue.length) {
        const nextQueue = []
        for (let i = 0; i < queue.length; i++) {
            const [ball, player] = queue[i]
            for (let j = 0; j < dirs.length; j++) {
                const dir = dirs[j]
                const reverseDir = [dir[0] ? -dir[0] : 0, dir[1] ? -dir[1] : 0]
                const nextBall = add(ball, dir)
                const nextPlayer = add(ball, reverseDir)
                const nextBallKey = getKey(nextBall)
                const nextPlayerKey = getKey(nextPlayer)
                if (
                    validate(nextBall) &&
                    validate(nextPlayer) &&
                    freeSet.has(nextBallKey) &&
                    freeSet.has(nextPlayerKey)
                ) {
                    const nextStateKey = getStateKey(nextBall, nextPlayer)
                    if (isConnected(player, nextPlayer, ball)) {
                        if (!stateCache.has(nextStateKey)) {
                            stateCache.add(nextStateKey)
                            if (nextBallKey === targetKey) {
                                return count
                            }
                            nextQueue.push([nextBall, nextPlayer])
                        }
                    }
                }
            }
        }
        queue = nextQueue
        count++
    }
    return -1
};